src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -2681,407 +1707,102 @@ void ObjectSynchronizer::audit_and_print_stats(bool on_exit) {
-      }
-    }
-  }
-  if (Atomic::load(&om_list_globals._wait_count) == chk_om_wait_count) {
-    out->print_cr("global_wait_count=%d equals chk_om_wait_count=%d",
-                  Atomic::load(&om_list_globals._wait_count), chk_om_wait_count);
+  if (l_in_use_count == ck_in_use_count) {
+    out->print_cr("in_use_count=" SIZE_FORMAT " equals ck_in_use_count="
+                  SIZE_FORMAT, l_in_use_count, ck_in_use_count);
   } else {
-    out->print_cr("ERROR: global_wait_count=%d is not equal to "
-                  "chk_om_wait_count=%d",
-                  Atomic::load(&om_list_globals._wait_count), chk_om_wait_count);
-    *error_cnt_p = *error_cnt_p + 1;
+    out->print_cr("WARNING: in_use_count=" SIZE_FORMAT " is not equal to "
+                  "ck_in_use_count=" SIZE_FORMAT, l_in_use_count,
+                  ck_in_use_count);
   }
-}
 
-// Check the global in-use list and count; log the results of the checks.
-void ObjectSynchronizer::chk_global_in_use_list_and_count(outputStream * out,
-                                                          int *error_cnt_p) {
-  int chk_om_in_use_count = 0;
-  ObjectMonitor* cur = NULL;
-  if ((cur = get_list_head_locked(&om_list_globals._in_use_list)) != NULL) {
-    // Marked the global in-use list head so process the list.
-    while (true) {
-      chk_in_use_entry(NULL /* jt */, cur, out, error_cnt_p);
-      chk_om_in_use_count++;
-
-      cur = lock_next_for_traversal(cur);
-      if (cur == NULL) {
-        break;
-      }
-    }
-  }
-  int l_in_use_count = Atomic::load(&om_list_globals._in_use_count);
-  if (l_in_use_count == chk_om_in_use_count) {
-    out->print_cr("global_in_use_count=%d equals chk_om_in_use_count=%d",
-                  l_in_use_count, chk_om_in_use_count);
+  size_t ck_in_use_max = _in_use_list.max();
+  if (l_in_use_max == ck_in_use_max) {
+    out->print_cr("in_use_max=" SIZE_FORMAT " equals ck_in_use_max="
+                  SIZE_FORMAT, l_in_use_max, ck_in_use_max);
   } else {
-    // With fine grained locks on the monitor lists, it is possible for
-    // an exiting JavaThread to put its in-use ObjectMonitors on the
-    // global in-use list after chk_om_in_use_count is calculated above.
-    out->print_cr("WARNING: global_in_use_count=%d is not equal to chk_om_in_use_count=%d",
-                  l_in_use_count, chk_om_in_use_count);
+    out->print_cr("WARNING: in_use_max=" SIZE_FORMAT " is not equal to "
+                  "ck_in_use_max=" SIZE_FORMAT, l_in_use_max, ck_in_use_max);
   }
 }
 
 // Check an in-use monitor entry; log any errors.
-void ObjectSynchronizer::chk_in_use_entry(JavaThread* jt, ObjectMonitor* n,
-                                          outputStream * out, int *error_cnt_p) {
+void ObjectSynchronizer::chk_in_use_entry(ObjectMonitor* n, outputStream* out,
+                                          int* error_cnt_p) {
+  if (n->owner_is_DEFLATER_MARKER()) {
+    // This should not happen, but if it does, it is not fatal.
```

Comment:
Deflating an in-use monitor is not fatal? Please explain how things would recover.

Reply:
When the MonitorDeflationThread is doing its final async deflation cycle before
VM shutdown, it is possible for it to finish a deflation pass on the in-use list and
then block for the final safepoint before unlinking the deflated ObjectMonitors.

If the VMThread happens to be doing an audit_and_print_stats() call when this
happens (see the end of SafepointSynchronize::do_cleanup_tasks()), then we
don't want that audit to FAIL.

As for recovery, if the VMThread is doing a final audit, then when it sees an
already deflated ObjectMonitor on the in-use list, it will take care of the unlinking
and deleting...